Use ThorVG instead of NanoSVG for importing SVGs

fire opened the pull request:
ThorVG is a platform-independent portable library for drawing vector-based scene and animation.

Alternative to https://github.com/godotengine/godot/pull/49601.

Resolves the light to dark for fill, stop colors and strokes.

https://github.com/Samsung/thorvg

Fixes: https://github.com/godotengine/godot/issues/49205
Fixes: https://github.com/godotengine/godot/issues/16180 (Due to increased support we are able to resolve SVG issues better).
Fixes: https://github.com/godotengine/godot/issues/33641 (Changing SVG libraries replaces old bugs with new bugs.)
Fixes: https://github.com/godotengine/godot/issues/41339 (Via https://github.com/Samsung/thorvg/pull/530)
Fixes: https://github.com/godotengine/godot/issues/51763
Fixes: https://github.com/godotengine/godot-proposals/issues/2912

Xrayez commented on modules/thorvg/README.md:
```diff
@@ -0,0 +1,26 @@
+# LunaSVG
```
Needs update, what how differs from LunaSVG?

```suggestion
# ThorSVG
```

fire replied:
Removed.

fire commented:
There's a high velocity in the thorvg project. I posted broken test cases and the team responded the day after.

https://github.com/Samsung/thorvg/issues/444

fire commented:
As requested by Groud:

https://github.com/godotengine/godot/commit/70c5c1b8810c0fbc98e374962b5e1a28c1112fe6 (thorvg)

godot.windows.opt.tools.64.exe 89.8 MB (94,245,888 bytes) uncompressed.

https://github.com/godotengine/godot/commit/a01ea509f5b2417cad1712f74642462fe85f5170 (Merge pull request #49826 from groud/fix_mesh_transform)

godot.windows.opt.tools.64.exe 89.7 MB (94,151,680 bytes) uncompressed.

akien-mga commented on editor/editor_themes.cpp:
```diff
@@ -37,8 +37,8 @@
 #include "modules/modules_enabled.gen.h"
-#ifdef MODULE_SVG_ENABLED
-#include "modules/svg/image_loader_svg.h"
+#ifdef MODULE_THORVG_ENABLED
+#include "modules/thorvg/image_loader_thorvg.h"
```
I suggest keeping the `modules/svg/image_loader_svg.h` name, only the thirdparty library needs to change (nanosvg -> thorvg).

That avoids having to do all those renames and the library change is transparent to the rest of the engine.

akien-mga commented on editor/editor_themes.cpp:
```diff
@@ -241,7 +241,7 @@ void editor_register_and_generate_icons(Ref<Theme> p_theme, bool p_dark_theme =
 	dark_icon_color_dictionary[Color::html("#dbab09")] = warning_color;
 
-	ImageLoaderSVG::set_convert_colors(&dark_icon_color_dictionary);
+	// ImageLoaderSVG::set_convert_colors(&dark_icon_color_dictionary);
```
This might need to be re-implemented? I haven't checked but I guess it's used to convert colors from dark theme to light theme?

fire replied:
We discussed in a thread with @Calinou that the SVG importer should not be replacing colors..

That the operation should be done on the xml file. Stuck here.

akien-mga commented on modules/thorvg/SCsub:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+import glob
```
This doesn't seem to be used.

akien-mga commented on modules/thorvg/SCsub:
```diff
@@ -0,0 +1,69 @@
+]
+thirdparty_sources = [thirdparty_dir + file for file in thor_src]
+env_thorvg.Prepend(CPPPATH=["."])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/lib/"])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/lib/sw_engine/"])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/loaders/raw/"])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/loaders/svg/"])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/loaders/png/"])
+env_thorvg.Prepend(CPPPATH=["#thirdparty/thorvg/src/loaders/jpg/"])
```
I'd be a bit surprised if all those explicit include paths are actually needed.

fire replied:
I had trouble removing them.

akien-mga commented on modules/thorvg/SCsub:
```diff
@@ -0,0 +1,69 @@
+    "src/lib/sw_engine/tvgSwStroke.cpp",
+]
+thirdparty_sources = [thirdparty_dir + file for file in thor_src]
+env_thorvg.Prepend(CPPPATH=["."])
```
I doubt this is necessary?

fire replied:
This is used to add the "godot\modules\svg\config.h" file.

akien-mga commented on modules/thorvg/image_loader_thorvg.cpp:
```diff
@@ -0,0 +1,114 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "image_loader_thorvg.h"
+#include "core/error/error_macros.h"
+#include "core/math/vector2.h"
+#include "thorvg.h"
+#include <stdint.h>
+#include <memory>
+
+#include "thirdparty/thorvg/inc/thorvg.h"
```
Please review https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

Proper style here would be:
```suggestion
#include "image_loader_thorvg.h"

#include "core/error/error_macros.h"
#include "core/math/vector2.h"

#include <memory>
#include <stdint.h>
#include <thorvg.h>
```

`thirdparty/thorvg/inc/thorvg.h` is redundant and shouldn't be needed if you get the CPPPATH right.

akien-mga commented on modules/thorvg/image_loader_thorvg.h:
```diff
@@ -28,46 +28,28 @@
-class SVGRasterizer {
-	NSVGrasterizer *rasterizer;
+#include "core/templates/local_vector.h"
+#include "thirdparty/thorvg/inc/thorvg.h"
```
This doesn't seem used in this file. We should only include what we use.

akien-mga commented on modules/thorvg/image_loader_thorvg.h:
```diff
@@ -28,46 +28,28 @@
-	NSVGrasterizer *rasterizer;
+#include "core/templates/local_vector.h"
+#include "thirdparty/thorvg/inc/thorvg.h"
+#include <stdint.h>
+#include <iostream>
+#include <memory>
+#include <sstream>
```
None of these are used? This is a very simple header, I wouldn't expect much more than `"core/io/image_loader.h"`.

akien-mga commented on thirdparty/README.md:
```diff
@@ -1,3 +1,4 @@
+
```
Unwarranted style change :)

akien-mga commented on thirdparty/README.md:
```diff
@@ -643,6 +644,20 @@ Files extracted from upstream source:
 
+## thorvg
+
+- Upstream: https://github.com/Samsung/thorvg.git
```
```suggestion
- Upstream: https://github.com/Samsung/thorvg
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -643,6 +644,20 @@ Files extracted from upstream source:
 - `tinyexr.{cc,h}`
 
 
+## thorvg
```
Remember to also remove the part about nanosvg since we remove it.

akien-mga commented:
I made fixes myself directly as it would have been too tedious to do through line based review comments. Please review eee0d1b to validate the fixups.

There's still a few technical issues to resolve before this can be merged:

- There's no color conversion anymore so the light theme is broken.
- The `p_upscale` parameter is no longer respected. I suspect this will break HiDPI upscaling.
- The instructions in `README.md` for copying the sources from upstream are incomplete. We're not copying the whole `src` folder as written, instead we seem to cherry-pick only what we need (which is good), but this needs to be described so that I don't need to reverse engineer it myself on updates.

akien-mga commented on modules/svg/SCsub:
```diff
@@ -44,22 +43,29 @@ thor_src = [
 module_obj = []
 
 env_svg.add_source_files(module_obj, "*.cpp")
-env_svg.disable_warnings()
```
You should not disable warnings on Godot code - that's why we always build thirdparty code in a dedicated cloned environment for itself.

fire replied:
This was an oversight.

akien-mga commented on modules/svg/config.h:
```diff
@@ -1,33 +0,0 @@
-/*************************************************************************/
-/*  config.h                                                             */
```
This was bogus. ThorVG does require a `config.h` but it's better to give it a proper one - I've done that in `thirdparty/thorvg/inc/config.h`.

akien-mga commented on modules/svg/image_loader_svg.cpp:
```diff
@@ -68,10 +70,12 @@ void ImageLoaderSVG::create_image_from_string(Ref<Image> p_image, String p_strin
 	picture->viewbox(nullptr, nullptr, &fw, &fh);
-	ERR_FAIL_COND(Math::is_zero_approx(p_scale));
+
+	// FIXME: p_upsample is not used.
```
Could be as simple as multiplying it to `fw * p_scale` and `fh * p_scale` below as done in the previous loader, but should be checked.

akien-mga commented on modules/svg/register_types.cpp:
```diff
@@ -45,5 +47,6 @@ void register_svg_types() {
 }
 
 void unregister_svg_types() {
+	memdelete(image_loader_svg);
```
I'm not sure why you removed this? If you allocate memory, you also need to free it, otherwise it's a leak.

fire replied:
Attending to this.

fire commented:
1. There's no color conversion anymore so the light theme is broken.

I was advised to do this outside of godot importer, but this is unsolved.

2. The p_upscale parameter is no longer respected. I suspect this will break HiDPI upscaling.

On it.

3. The instructions in README.md for copying the sources from upstream are incomplete. We're not copying the whole src folder as written, instead we seem to cherry-pick only what we need (which is good), but this needs to be described so that I don't need to reverse engineer it myself on updates.

I remember the steps were delete .meson and delete a list of folders. Will make a script to catalog the changes.

fire commented:
Reviewers please check the `editor icon conversion code`.`

fire commented:
Todo:

- [ ] Review the includes (Does anyone have the guide?)
- [ ] Currently the code is forced single threaded. Release requirement?
- [ ] Second review

mhilbrunner commented:
PR review meeting note: looks good, needs an in-depth review, we should compare performance with the current state

fire committed:
Use ThorVG instead of NanoSVG for importing SVGs

ThorVG is a platform-independent portable library for drawing vector-based
scene and animation.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

akien-mga commented on modules/svg/register_types.cpp:
```diff
@@ -30,10 +30,8 @@
-#include "core/os/os.h"
 #include "image_loader_svg.h"
 
-#include "thirdparty/thorvg/inc/config.h"
```
Normally a library is responsible for including its own `config.h` where relevant, we should haven't to do it ourselves before including the main entry point. Should be double checked in case ThorVG doesn't do this properly. It seems to include it itself in two places:
```
src/lib/tvgLzw.cpp
56:#include "config.h"

src/lib/tvgCommon.h
25:#include "config.h"
```

fire replied:
Seems work removed in modules/svg/register_types.cpp.

akien-mga commented:
Did some speed tests to assess if the new importer is slower or faster than nanosvg, by judging the impact on the editor start time:

### Dark theme (default, no conversion)

```
$ hyperfine -iw1 "bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit" "bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit
  Time (mean ± σ):      8.598 s ±  0.812 s    [User: 7.619 s, System: 0.376 s]
  Range (min … max):    7.902 s … 10.174 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Benchmark #2: bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit
  Time (mean ± σ):      8.781 s ±  0.505 s    [User: 7.825 s, System: 0.383 s]
  Range (min … max):    8.137 s …  9.785 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Summary
  'bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit' ran
    1.02 ± 0.11 times faster than 'bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit'
```

`.docthread` is just a build from a couple of days ago with #50864, it was merged so it should perform similarly to vanilla `master`.

So the version with ThorVG might be a tiny bit slower, but it's not really significant from my results (too much variance).
I'd say it's pretty much similar.

### Light theme (color conversion)

```
$ hyperfine -iw1 "bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit" "bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit
  Time (mean ± σ):      8.781 s ±  0.307 s    [User: 7.356 s, System: 0.357 s]
  Range (min … max):    8.328 s …  9.215 s    10 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark #2: bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit
  Time (mean ± σ):     11.069 s ±  0.093 s    [User: 9.509 s, System: 0.353 s]
  Range (min … max):   11.009 s … 11.305 s    10 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'bin/godot.linuxbsd.tools.64.docthread ~/tmp/godot/projects/FreeFall/project.godot --quit' ran
    1.26 ± 0.05 times faster than 'bin/godot.linuxbsd.tools.64 ~/tmp/godot/projects/FreeFall/project.godot --quit'
```

Here it shows on the other hand that the color conversion takes more than 2 s when compared to the timing with dark theme, while the previous method with nanosvg was not significantly slower.

That can also be seen by changing the theme from Dark the Light in the editor, it takes multiple seconds before it's actually applied (it used to be a bit slow, but not as much).

There might be room for optimizing the way the color conversions are done in this PR. Could be left for a follow-up but we should keep track of it in an issue, for Light Theme users a startup time of 11 s when it used to be less than 9 s is a big difference.

Maybe @Calinou can get better benchmark results on a more stable system (my laptop tends to do whatever it feels like with CPU clocks).

fire commented:
At the computer now, will look at the new commits and optimize the color conversion (it is slow).

fire commented:
Is there a way in Godot Engine to take an array of string searches and an array of string replacements and does the substitution once rather than per number of searches? Used to convert the svg themes.

fire commented:
@akien-mga I think we're ready and that it would be appropriate for this pr to be merge and wait for a follow-up but tracked in an issue..

> For Light Theme users a startup time of 11 s when it used to be less than 9 s is a big difference.

Calinou commented:
Benchmark with `target=release_debug use_lto=no` editor build:

**OS:** Fedora 34 (KDE, X11)
**CPU:** Intel Core i7-6700K
**GPU:** GeForce GTX 1080

`godot` includes this PR rebased against d13c3c929. `godot.vanilla` is d13c3c929 and doesn't include this PR.

```
❯ hyperfine -w1 -i "godot /tmp/4/project.godot --quit" "godot.vanilla /tmp/4/project.godot --quit"
Benchmark #1: godot /tmp/4/project.godot --quit
  Time (mean ± σ):      3.617 s ±  0.355 s    [User: 1.852 s, System: 0.245 s]
  Range (min … max):    3.092 s …  3.868 s    10 runs
 
Benchmark #2: godot.vanilla /tmp/4/project.godot --quit
  Time (mean ± σ):      3.275 s ±  0.357 s    [User: 1.821 s, System: 0.249 s]
  Range (min … max):    2.916 s …  3.816 s    10 runs
 
Summary
  'godot.vanilla /tmp/4/project.godot --quit' ran
    1.10 ± 0.16 times faster than 'godot /tmp/4/project.godot --quit'

~/.local/bin 1m 18s
❯ hyperfine -w1 -i "godot /tmp/4/project.godot --quit" "godot.vanilla /tmp/4/project.godot --quit"
Benchmark #1: godot /tmp/4/project.godot --quit
  Time (mean ± σ):      3.923 s ±  0.404 s    [User: 2.268 s, System: 0.244 s]
  Range (min … max):    3.332 s …  4.264 s    10 runs
 
Benchmark #2: godot.vanilla /tmp/4/project.godot --quit
  Time (mean ± σ):      3.202 s ±  0.313 s    [User: 1.804 s, System: 0.253 s]
  Range (min … max):    2.927 s …  3.787 s    10 runs
 
Summary
  'godot.vanilla /tmp/4/project.godot --quit' ran
    1.23 ± 0.17 times faster than 'godot /tmp/4/project.godot --quit'
```

fire commented:
Modified with a new code for replacing colors. The code subjectively feels 2 s versus 1.5 s in master to change the theme.

akien-mga commented on modules/svg/image_loader_svg.cpp:
```diff
@@ -30,132 +30,120 @@
-				Color new_color = n_c;
-				replace_colors.old_colors.push_back(old_color.to_abgr32());
-				replace_colors.new_colors.push_back(new_color.to_abgr32());
+		::Color c = ::Color::html(temp_color);
```
Does your IDE add this scoping? I don't see why it's needed here.

akien-mga replied:
I'll fix it while doing some testing, I think the color replacement method could still be simplified, we have plenty of methods in `String` to do search and replace, doing it char by char seems pretty convoluted.

fire replied:
My clang compile requires the "::" scoping.

akien-mga replied:
What clang is that? Works fine for me and on CI (macOS + iOS + Android all use clang).

We use `Color` everywhere in the codebase. What does it conflict with here?

akien-mga commented:
Just noting down some LTO warnings from ThorVG to report upstream (after checking that they're still reproducible in their `master` branch):
```
thirdparty/thorvg/src/loaders/external_png/tvgPngLoader.h:27:7: warning: type 'struct PngLoader' violates the C++ One Definition Rule [-Wodr]
   27 | class PngLoader : public LoadModule
      |       ^
thirdparty/thorvg/src/loaders/png/tvgPngLoader.h:29:7: note: a type with the same name but different number of polymorphic bases is defined in another translation unit
   29 | class PngLoader : public LoadModule, public Task
      |       ^
thirdparty/thorvg/src/lib/tvgTaskScheduler.h:42:8: note: the extra base is defined here
   42 | struct Task
      |        ^
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:131:8: warning: type 'struct LodePNGInfo' violates the C++ One Definition Rule [-Wodr]
  131 | struct LodePNGInfo
      |        ^
thirdparty/basis_universal/encoder/lodepng.h:412: note: a different type is defined in another translation unit
  412 | typedef struct LodePNGInfo {
      | 
thirdparty/basis_universal/encoder/lodepng.h:443: note: the first difference of corresponding definitions is field 'background_defined'
  443 |   unsigned background_defined; /*is a suggested background color given?*/
      | 
thirdparty/basis_universal/encoder/lodepng.h:412: note: a type with different number of fields is defined in another translation unit
  412 | typedef struct LodePNGInfo {
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:70:8: warning: type 'struct LodePNGDecompressSettings' violates the C++ One Definition Rule [-Wodr]
   70 | struct LodePNGDecompressSettings
      |        ^
thirdparty/basis_universal/encoder/lodepng.h:262: note: a different type is defined in another translation unit
  262 | struct LodePNGDecompressSettings {
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:74:14: note: the first difference of corresponding definitions is field 'ignore_nlen'
   74 |     unsigned ignore_nlen; /*ignore complement of len checksum in uncompressed blocks*/
      |              ^
thirdparty/basis_universal/encoder/lodepng.h:267: note: a field with different name is defined in another translation unit
  267 |   unsigned (*custom_zlib)(unsigned char**, size_t*,
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:144:8: warning: type 'struct LodePNGDecoderSettings' violates the C++ One Definition Rule [-Wodr]
  144 | struct LodePNGDecoderSettings
      |        ^
thirdparty/basis_universal/encoder/lodepng.h:618: note: a different type is defined in another translation unit
  618 | typedef struct LodePNGDecoderSettings {
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:146:31: note: the first difference of corresponding definitions is field 'zlibsettings'
  146 |     LodePNGDecompressSettings zlibsettings; /*in here is the setting to ignore Adler32 checksums*/
      |                               ^
thirdparty/basis_universal/encoder/lodepng.h:619: note: a field of same name but different type is defined in another translation unit
  619 |   LodePNGDecompressSettings zlibsettings; /*in here is the setting to ignore Adler32 checksums*/
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:70:8: note: type 'struct LodePNGDecompressSettings' itself violates the C++ One Definition Rule
   70 | struct LodePNGDecompressSettings
      |        ^
thirdparty/basis_universal/encoder/lodepng.h:730: warning: type 'struct LodePNGState' violates the C++ One Definition Rule [-Wodr]
  730 | typedef struct LodePNGState {
      | 
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:161:8: note: a type with different virtual table pointers is defined in another translation unit
  161 | struct LodePNGState
      |        ^
thirdparty/thorvg/src/loaders/png/tvgPngLoader.h:41:5: warning: type of '__ct_comp ' does not match original declaration [-Wlto-type-mismatch]
   41 |     PngLoader();
      |     ^
thirdparty/thorvg/src/loaders/external_png/tvgPngLoader.cpp:45:1: note: '__ct_comp ' was previously declared here
   45 | PngLoader::PngLoader()
      | ^
thirdparty/thorvg/src/loaders/png/tvgLodePng.h:54: warning: type 'LodePNGColorType' violates the C++ One Definition Rule [-Wodr]
   54 | enum LodePNGColorType
      | 
thirdparty/basis_universal/encoder/lodepng.h:99: note: an enum with different number of values is defined in another translation unit
   99 | typedef enum LodePNGColorType {
      |
```
Some are internal redefinitions in ThorVG itself, and others are conflicts between the vendored LodePNG versions from Basis Universal and ThorVG.

akien-mga commented:
Tested current PR with `scons platform=linuxbsd tools=yes target=release_debug use_lto=yes` and the results are pretty good.

## Dark theme

```
$ hyperfine -iw1 "bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit" "bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      3.889 s ±  0.412 s    [User: 2.368 s, System: 0.289 s]
  Range (min … max):    3.424 s …  4.480 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      3.866 s ±  0.494 s    [User: 2.360 s, System: 0.288 s]
  Range (min … max):    3.116 s …  4.431 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit' ran
    1.01 ± 0.17 times faster than 'bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit'
```

## Light theme

```
$ hyperfine -iw1 "bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit" "bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      4.009 s ±  0.343 s    [User: 2.375 s, System: 0.287 s]
  Range (min … max):    3.514 s …  4.430 s    10 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: The first benchmarking run for this command was significantly slower than the rest (4.430 s). This could be caused by (filesystem) caches that were not filled until after the first run. You should consider using the '--warmup' option to fill those caches before the actual benchmark. Alternatively, use the '--prepare' option to clear the caches before each timing run.
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      3.978 s ±  0.424 s    [User: 2.360 s, System: 0.286 s]
  Range (min … max):    3.492 s …  4.446 s    10 runs
 
  Warning: Ignoring non-zero exit code.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.thorvg ~/tmp/godot/test/NewTest/project.godot --quit' ran
    1.01 ± 0.14 times faster than 'bin/godot.linuxbsd.opt.tools.64.vanilla ~/tmp/godot/test/NewTest/project.godot --quit'
```

So it's pretty much the same for both dark and light theme with both current `master` (9b3535a33a1dda541a3a39e7786b8428fadbff6c) and this PR.

akien-mga commented:
I pushed an update that simplifies the color conversion code by using existing `String` methods to do substring replacement, and by storing HTML code as strings in the conversion dictionary instead of doing an unnecessary HTML -> Color -> HTML conversion.

Despite the theoretical micro optimization, that seemed to make no difference at all in color conversion speed. But the code is nicer IMO :)

akien-mga commented:
I need to do some more testing, I noticed that some assumptions don't work, as some files contain 3-letter color codes like `fill="#fff"`.

akien-mga commented:
Alright, I think it's ready.

I tweaked the color replacement code to make it handle the whole quoted string instead of hardcoding support for only the `"#abcdef` format. So now it should support:
- Color codes without `#` (assuming they're supported in the spec, didn't check)
- 3-letter color codes (`#abc`)
- Color codes with alpha (4 and 8 letter)
- Named colors supported by Godot (`black`, `blue`, etc.)

While developing, I made it print colors for which it didn't find a conversion, so we can review and assess if some of those are meant to be converted:

```
Couldn't find color '#000' for property 'stroke="'.
Couldn't find color '#003e7a' for property 'fill="'.
Couldn't find color '#00f' for property 'fill="'.
Couldn't find color '#044b94' for property 'fill="'.
Couldn't find color '#0f0' for property 'fill="'.
Couldn't find color '#1aab1a' for property 'fill="'.
Couldn't find color '#288027' for property 'stop-color="'.
Couldn't find color '#2998ff' for property 'fill="'.
Couldn't find color '#3552b1' for property 'fill="'.
Couldn't find color '#404040' for property 'stroke="'.
Couldn't find color '#414042' for property 'fill="'.
Couldn't find color '#41562e' for property 'stroke="'.
Couldn't find color '#424242' for property 'stroke="'.
Couldn't find color '#4646ff' for property 'fill="'.
Couldn't find color '#46ff46' for property 'fill="'.
Couldn't find color '#474747' for property 'fill="'.
Couldn't find color '#478cbf' for property 'fill="'.
Couldn't find color '#4b4b4b' for property 'fill="'.
Couldn't find color '#62aeff' for property 'stop-color="'.
Couldn't find color '#66ff9e' for property 'stop-color="'.
Couldn't find color '#68b6ff' for property 'fill="'.
Couldn't find color '#68b6ff' for property 'stop-color="'.
Couldn't find color '#68b6ff' for property 'stroke="'.
Couldn't find color '#68d0ea' for property 'fill="'.
Couldn't find color '#699ce8' for property 'fill="'.
Couldn't find color '#69ec9e' for property 'fill="'.
Couldn't find color '#69f' for property 'fill="'.
Couldn't find color '#6e6e6e' for property 'fill="'.
Couldn't find color '#70bfff' for property 'fill="'.
Couldn't find color '#75d1e6' for property 'stop-color="'.
Couldn't find color '#77ce57' for property 'fill="'.
Couldn't find color '#7aff70' for property 'fill="'.
Couldn't find color '#808080' for property 'fill="'.
Couldn't find color '#84ffee' for property 'stop-color="'.
Couldn't find color '#919191' for property 'fill="'.
Couldn't find color '#999' for property 'fill="'.
Couldn't find color '#a2d2ff' for property 'fill="'.
Couldn't find color '#a2d2ff' for property 'stop-color="'.
Couldn't find color '#a3b6f3' for property 'fill="'.
Couldn't find color '#a5b7f3' for property 'fill="'.
Couldn't find color '#a5b7f4' for property 'fill="'.
Couldn't find color '#a5b8f3' for property 'fill="'.
Couldn't find color '#afafaf' for property 'fill="'.
Couldn't find color '#b3b3b3' for property 'fill="'.
Couldn't find color '#b8ea68' for property 'fill="'.
Couldn't find color '#b8ea68' for property 'stop-color="'.
Couldn't find color '#b8ea68' for property 'stroke="'.
Couldn't find color '#ccc' for property 'fill="'.
Couldn't find color '#ccc' for property 'stop-color="'.
Couldn't find color '#cf68ea' for property 'fill="'.
Couldn't find color '#cf68ea' for property 'stop-color="'.
Couldn't find color '#d5d5d5' for property 'fill="'.
Couldn't find color '#d6d6d6' for property 'fill="'.
Couldn't find color '#d9d9d9' for property 'fill="'.
Couldn't find color '#dbee15' for property 'stop-color="'.
Couldn't find color '#e1e1e1' for property 'fill="'.
Couldn't find color '#e6e6e6' for property 'fill="'.
Couldn't find color '#ea686c' for property 'fill="'.
Couldn't find color '#ea686c' for property 'stop-color="'.
Couldn't find color '#ea686c' for property 'stroke="'.
Couldn't find color '#eac968' for property 'fill="'.
Couldn't find color '#eae068' for property 'fill="'.
Couldn't find color '#eec315' for property 'stop-color="'.
Couldn't find color '#f2f2f2' for property 'fill="'.
Couldn't find color '#f5f5f5' for property 'fill="'.
Couldn't find color '#f5f5f5' for property 'stroke="'.
Couldn't find color '#f6f6f6' for property 'fill="'.
Couldn't find color '#f6f6f6' for property 'stroke="'.
Couldn't find color '#f70000' for property 'stop-color="'.
Couldn't find color '#f7f5cf' for property 'fill="'.
Couldn't find color '#fefefe' for property 'fill="'.
Couldn't find color '#fefeff' for property 'fill="'.
Couldn't find color '#ff4646' for property 'fill="'.
Couldn't find color '#ff5d5d' for property 'fill="'.
Couldn't find color '#ff7070' for property 'fill="'.
Couldn't find color '#ff7a7a' for property 'stop-color="'.
Couldn't find color '#ff8585' for property 'fill="'.
Couldn't find color '#ffa62a' for property 'fill="'.
```

akien-mga commented:
Thanks!
